target/linux/mpc85xx/files/arch/powerpc/boot/dts/fsl/p2020rdb.dts

```diff
@@ -0,0 +1,318 @@
+/*
+ * P2020 RDB Device Tree Source
+ *
+ * Copyright 2009-2012 Freescale Semiconductor Inc.
+ * Copyright 2018      Pawel Dembicki <paweldembicki@gmail.com>
+ *
+ * This program is free software; you can redistribute  it and/or modify it
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/include/ "p2020si-pre.dtsi"
+
+/ {
+	model = "Freescale P2020RDB";
+	compatible = "fsl,P2020RDB";
+
+	aliases {
+		ethernet0 = &enet0;
+		ethernet1 = &enet1;
+		ethernet2 = &enet2;
+		serial0 = &serial0;
+		serial1 = &serial1;
+		pci0 = &pci0;
+		pci1 = &pci1;
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		rst_phy_eth2 {
+			gpio-export,name = "rst_phy_eth2";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	memory {
+		device_type = "memory";
+	};
+
+	lbc: localbus@ffe05000 {
+		reg = <0 0xffe05000 0 0x1000>;
+
+		/* NOR and NAND Flashes */
+		ranges = <0x0 0x0 0x0 0xef000000 0x01000000
+			  0x1 0x0 0x0 0xffa00000 0x00040000
+			  0x2 0x0 0x0 0xffb00000 0x00020000>;
+
+		nor@0,0 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			compatible = "cfi-flash";
+			reg = <0x0 0x0 0x1000000>;
+			bank-width = <2>;
+			device-width = <1>;
+
+			partition@0 {
+				/* This location must not be altered  */
+				/* 256KB for Vitesse 7385 Switch firmware */
+				reg = <0x0 0x00040000>;
+				label = "NOR (RO) Vitesse-7385 Firmware";
+				read-only;
+			};
+
+			partition@40000 {
+				/* 256KB for DTB Image */
+				reg = <0x00040000 0x00040000>;
+				label = "dtb";
+			};
+
+			partition@80000 {
+				/* 3.5 MB for Linux Kernel Image */
+				reg = <0x00080000 0x00380000>;
+				label = "kernel";
+			};
+
+			partition@400000 {
+				/* 11MB for JFFS2 based Root file System */
+				reg = <0x00400000 0x00b00000>;
+				label = "rootfs";
+			};
+
+			partition@f00000 {
+				/* This location must not be altered  */
+				/* 512KB for u-boot Bootloader Image */
+				/* 512KB for u-boot Environment Variables */
+				reg = <0x00f00000 0x00100000>;
+				label = "u-boot";
+				read-only;
+			};
+			firmware@40000 {
+				reg = <0x00040000 0x00ec0000>;
+				label = "firmware";
```

Comment:
Usually, you only need to specify the label OR the node-name. The sysupgrade and the mtd-utils should be smart enough to figure it out. If not something is broken.

Reply:
The same name convention is used in all dts in this target. I can't change node name to "partition", because it causes "Duplicate node name" error.

Reply:
> The same name convention is used in all dts in this target. I can't change node name to "partition", because it causes "Duplicate node name" error.

Which shows that you are doing it wrong :-).

Why not remove the partitions at 0x40000, 0x80000, 0x400000 and add the factory partition covering the whole space instead?

Reply:
I reworked partition list. 

@mkresin 
I think about use "denx,uimage" partition spliter. But at start firmware partition is dtb area. Could You recomend method for this problem? Add offset property to spliter?

Reply:
> I think about use "denx,uimage" partition spliter. But at start firmware partition is dtb area. Could You recomend method for this problem? Add offset property to spliter? 

If you want to go that route you should check if your u-boot might already support FIT.
If not, you could still try the legacy multi-file uimage approach. In both cases the dtb (+ ramdisk if you have any) can be part of the fit/uimage. 

If you are looking for pointers for the multi-file images checkout the apm821xx. I'm using it for initramfs for the WNDR4700, MR24, MX60, MBL and for the WNDAP620/660s:
https://github.com/openwrt/openwrt/blob/master/target/linux/apm821xx/image/Makefile#L72

Reply:
At this moment device won't booting from FIT image. Only imxtract both images and then regular boot work. 

```
=> imxtract 02000000 kernel@1 02400000
## Copying 'kernel@1' subimage from FIT image at 02000000 ...
crc32+ sha1+    Uncompressing part 0 ... OK
=> imxtract 02000000 fdt@1 c00000
## Copying 'fdt@1' subimage from FIT image at 02000000 ...
crc32+ sha1+    Loading part 253 ... OK
=> bootm 02400000 - c00000
WARNING: adjusting available memory to 30000000
## Booting kernel from Legacy Image at 02400000 ...
   Image Name:   POWERPC OpenWrt Linux-4.19.9
   Created:      2019-01-01  18:33:32 UTC
   Image Type:   PowerPC Linux Kernel Image (gzip compressed)
   Data Size:    2899371 Bytes = 2.8 MiB
   Load Address: 00000000
   Entry Point:  00000000
   Verifying Checksum ... OK
## Flattened Device Tree blob at 00c00000
   Booting using the fdt blob at 0x00c00000
   Uncompressing Kernel Image ... OK
   Loading Device Tree to 03ffa000, end 03ffffaf ... OK
WARNING: could not find compatible node fsl-usb2-dr: FDT_ERR_NOTFOUND.
[    0.000000] Memory CAM mapping: 256/256/256 Mb, residual: 256Mb
[    0.000000] Linux version 4.19.9 (krolik@ubuntu-VirtualBox) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r8859+56-f36bc3f9b1)) #0 SMP Tue Jan 1 15:23:09 2019
[    0.000000] Using P2020 RDB machine description
```

```
=> iminfo

## Checking Image at 02000000 ...
   FIT image found
   FIT description: POWERPC OpenWrt FIT (Flattened Image Tree)
   Created:         2019-01-01  18:33:32 UTC
    Image 0 (kernel@1)
     Description:  POWERPC OpenWrt Linux-4.19.9
     Type:         Kernel Image
     Compression:  gzip compressed
     Data Start:   0x020000ec
     Data Size:    2897673 Bytes = 2.8 MiB
     Architecture: PowerPC
     OS:           Linux
     Load Address: 0x00000000
     Entry Point:  0x00000000
     Hash algo:    crc32
     Hash value:   2b83681a
     Hash algo:    sha1
     Hash value:   0d3ea6a1d0400782a2cfa1e16ea8b18d28a50e03
    Image 1 (fdt@1)
     Description:  POWERPC OpenWrt p2020rdb device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x022c3934
     Data Size:    12208 Bytes = 11.9 KiB
     Architecture: PowerPC
     Hash algo:    crc32
     Hash value:   268aea01
     Hash algo:    sha1
     Hash value:   cd2de619ed19b2e32246d46c322f2a6a8374d97b
    Default Configuration: 'config@1'
    Configuration 0 (config@1)
     Description:  OpenWrt
     Kernel:       kernel@1
     FDT:          fdt@1
## Checking hash(es) for FIT Image at 02000000 ...
   Hash(es) for Image 0 (kernel@1): crc32+ sha1+
   Hash(es) for Image 1 (fdt@1): crc32+ sha1+
=> bootm
WARNING: adjusting available memory to 30000000
## Booting kernel from FIT Image at 02000000 ...
   Using 'config@1' configuration
   Trying 'kernel@1' kernel subimage
     Description:  POWERPC OpenWrt Linux-4.19.9
     Type:         Kernel Image
     Compression:  gzip compressed
     Data Start:   0x020000ec
     Data Size:    2897673 Bytes = 2.8 MiB
     Architecture: PowerPC
     OS:           Linux
     Load Address: 0x00000000
     Entry Point:  0x00000000
     Hash algo:    crc32
     Hash value:   2b83681a
     Hash algo:    sha1
     Hash value:   0d3ea6a1d0400782a2cfa1e16ea8b18d28a50e03
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Flattened Device Tree from FIT Image at 02000000
   Using 'config@1' configuration
   Trying 'fdt@1' FDT blob subimage
     Description:  POWERPC OpenWrt p2020rdb device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x022c3934
     Data Size:    12208 Bytes = 11.9 KiB
     Architecture: PowerPC
     Hash algo:    crc32
     Hash value:   268aea01
     Hash algo:    sha1
     Hash value:   cd2de619ed19b2e32246d46c322f2a6a8374d97b
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x022c3934
   Uncompressing Kernel Image ... OK
   Loading Device Tree to 03ffa000, end 03ffffaf ... OK
WARNING: could not find compatible node fsl-usb2-dr: FDT_ERR_NOTFOUND.


```

I consider leave it as is.

Reply:
> At this moment device won't booting from FIT image

it's very close though. I've seen errors like this before. In my cases this was all due to the fdt being located at a unmappable address (either due to reserve-memory or conflict with the early kernel memory mapping) and it is possible to resolve it through either clearing the fdt_high env or setting it to a different value (for example you use 0xc00000).  

You could also try to see what happens if you enable 

CONFIG_PPC_EARLY_DEBUG


As for the "multi-file images", any luck with that?

Reply:
```
[    0.649807] 3 fixed-partitions partitions found on MTD device ef000000.nor
[    0.656746] Creating 3 MTD partitions on "ef000000.nor":
[    0.662065] 0x000000000000-0x000000040000 : "NOR (RO) Vitesse-7385 Firmware"
[    0.669606] 0x000000040000-0x000000f00000 : "firmware"
[    0.675217] 2 fit-fw partitions found on MTD device firmware
[    0.680880] Creating 2 MTD partitions on "firmware":
[    0.685843] 0x000000000000-0x0000002c0000 : "kernel"
[    0.691201] 0x0000002a4664-0x000000ec0000 : "rootfs"
[    0.696557] mtd: device 3 (rootfs) set to be root filesystem
[    0.702262] 1 squashfs-split partitions found on MTD device rootfs
[    0.708443] 0x0000004c0000-0x000000ec0000 : "rootfs_data"
[    0.714246] 0x000000f00000-0x000001000000 : "u-boot"
```

Success. PR updated.

Reply:
nice :+1: . Can you tell us what exactly problem was with FIT on the first time around?

Reply:
I'm not exactly sure.